concurrency/LogAggregationService.java

```diff
@@ -0,0 +1,492 @@
+  private final DeletionService deletionService;
+  private final Dispatcher dispatcher;
+
+  private LocalDirsHandlerService dirsHandler;
+  Path remoteRootLogDir;
+  String remoteRootLogDirSuffix;
+  private NodeId nodeId;
+
+  private final ConcurrentMap<ApplicationId, AppLogAggregator> appLogAggregators;
+
+  private final ExecutorService threadPool;
+  
+  public LogAggregationService(Dispatcher dispatcher, Context context,
+      DeletionService deletionService, LocalDirsHandlerService dirsHandler) {
+    super(LogAggregationService.class.getName());
+    this.dispatcher = dispatcher;
+    this.context = context;
+    this.deletionService = deletionService;
+    this.dirsHandler = dirsHandler;
+    this.appLogAggregators =
+        new ConcurrentHashMap<ApplicationId, AppLogAggregator>();
+    this.threadPool = Executors.newCachedThreadPool(
+        new ThreadFactoryBuilder()
+          .setNameFormat("LogAggregationService #%d")
+          .build());
+  }
+
+  protected void serviceInit(Configuration conf) throws Exception {
+    this.remoteRootLogDir =
+        new Path(conf.get(YarnConfiguration.NM_REMOTE_APP_LOG_DIR,
+            YarnConfiguration.DEFAULT_NM_REMOTE_APP_LOG_DIR));
+    this.remoteRootLogDirSuffix =
+        conf.get(YarnConfiguration.NM_REMOTE_APP_LOG_DIR_SUFFIX,
+            YarnConfiguration.DEFAULT_NM_REMOTE_APP_LOG_DIR_SUFFIX);
+
+    super.serviceInit(conf);
+  }
+
+  @Override
+  protected void serviceStart() throws Exception {
+    // NodeId is only available during start, the following cannot be moved
+    // anywhere else.
+    this.nodeId = this.context.getNodeId();
+    super.serviceStart();
+  }
+  
+  @Override
+  protected void serviceStop() throws Exception {
+    LOG.info(this.getName() + " waiting for pending aggregation during exit");
+    stopAggregators();
+    super.serviceStop();
+  }
+   
+  private void stopAggregators() {
+    threadPool.shutdown();
+    boolean supervised = getConfig().getBoolean(
+        YarnConfiguration.NM_RECOVERY_SUPERVISED,
+        YarnConfiguration.DEFAULT_NM_RECOVERY_SUPERVISED);
+    // if recovery on restart is supported then leave outstanding aggregations
+    // to the next restart
+    boolean shouldAbort = context.getNMStateStore().canRecover()
+        && !context.getDecommissioned() && supervised;
+    // politely ask to finish
+    for (AppLogAggregator aggregator : appLogAggregators.values()) {
+      if (shouldAbort) {
+        aggregator.abortLogAggregation();
+      } else {
+        aggregator.finishLogAggregation();
+      }
+    }
+    while (!threadPool.isTerminated()) { // wait for all threads to finish
+      for (ApplicationId appId : appLogAggregators.keySet()) {
+        LOG.info("Waiting for aggregation to complete for " + appId);
+      }
+      try {
+        if (!threadPool.awaitTermination(30, TimeUnit.SECONDS)) {
+          threadPool.shutdownNow(); // send interrupt to hurry them along
+        }
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)